samples/apps/copilot-chat-app/webapi/CopilotChat/Skills/ChatSkills/ChatSkill.cs

```diff
@@ -471,178 +245,233 @@ static SKContext RemoveExceptionFromContext(SKContext chatContext)
     {
-        try
+        // 1. Extract user intent from the conversation history.
+        var userIntent = await this.GetUserIntentAsync(chatContext);
+        if (chatContext.ErrorOccurred)
         {
-            JsonNode? jsonNode = JsonNode.Parse(openApiSkillResponse);
-            string contentType = jsonNode?["contentType"]?.ToString() ?? string.Empty;
-            if (contentType.StartsWith("application/json", StringComparison.InvariantCultureIgnoreCase))
-            {
-                var content = jsonNode?["content"]?.ToString() ?? string.Empty;
-                if (!string.IsNullOrWhiteSpace(content))
-                {
-                    json = content;
-                    return true;
-                }
-            }
-        }
-        catch (JsonException)
-        {
-            this._logger.LogDebug("Unable to extract JSON from planner response, it is likely not from an OpenAPI skill.");
-        }
-        catch (InvalidOperationException)
-        {
-            this._logger.LogDebug("Unable to extract JSON from planner response, it may already be proper JSON.");
+            return string.Empty;
         }
 
-        json = string.Empty;
-        return false;
-    }
-
-    /// <summary>
-    /// Try to optimize json from the planner response
-    /// based on token limit
-    /// </summary>
-    private string OptimizeOpenApiSkillJson(string jsonContent, int tokenLimit, Plan plan)
-    {
-        int jsonTokenLimit = (int)(tokenLimit * this._promptOptions.RelatedInformationContextWeight);
-
-        // Remove all new line characters + leading and trailing white space
-        jsonContent = Regex.Replace(jsonContent.Trim(), @"[\n\r]", string.Empty);
-        var document = JsonDocument.Parse(jsonContent);
-        string lastSkillInvoked = plan.Steps[^1].SkillName;
-        string lastSkillFunctionInvoked = plan.Steps[^1].Name;
-        bool trimSkillResponse = false;
+        // 2. Calculate the remaining token budget.
+        var remainingToken = this.GetChatContextTokenLimit(userIntent);
 
-        // The json will be deserialized based on the response type of the particular operation that was last invoked by the planner
-        // The response type can be a custom trimmed down json structure, which is useful in staying within the token limit
-        Type skillResponseType = this.GetOpenApiSkillResponseType(ref document, ref lastSkillInvoked, ref lastSkillFunctionInvoked, ref trimSkillResponse);
+        // 3. Acquire external information from planner
+        var externalInformationTokenLimit = (int)(remainingToken * this._promptOptions.ExternalInformationContextWeight);
+        var plan = await this.AcquireExternalInformationAsync(chatContext, userIntent, externalInformationTokenLimit);
 
-        if (trimSkillResponse)
+        // If plan is suggested, send back to user for approval before running
+        if (this._externalInformationSkill.ProposedPlan != null)
```

Comment:
It might make more sense to return the plan as the result of `AcquireExternalInformationAsync` then check the type of the response, i.e., GetType() == Plan to avoid having a class property tracking this

Right now, this way is fine because we instantiate a new kernel on every skill, but in case something changes like the kernel becomes persistent, it could introduce issues

Reply:
I believe `AcquireExternalInformationAsync` does return a plan.

Reply:
`AcquireExternalInformationAsync ` is a SK function, I don't think we can have an out param.

Reply:
Not blocking - we can handle this in a later PR